Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -766,40 +768,88 @@ extension BrowserViewController: SearchViewControllerDelegate {
+    override func canBePreventedByGestureRecognizer(preventingGestureRecognizer: UIGestureRecognizer!) -> Bool {
+        if preventingGestureRecognizer.description.rangeOfString("UIWebTouchEventsGestureRecognizer") != nil {
+            // We normally wouldn't want to allow touch events a chance to prevent this, but doing so can result.
+            // in lock ups of the entire UI.
+            if preventingGestureRecognizer.state == .Ended {
+                return true
             }
+            return false;
+        }
+        return preventingGestureRecognizer.description.rangeOfString("UIScrollViewPanGestureRecognizer") == nil
+    }
+
+    override init(target: AnyObject, action: Selector) {
+        super.init(target: target, action: action)
+    }
 
-            self.previousScroll = scrollView.contentOffset
+    override func touchesBegan(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if touches.count == 1 {
+            if let touch = touches.first as? UITouch {
+                previousScroll = touch.locationInView(bvc?.view)
+            }
+        } else {
+            // If a second finer comes down, we'll stop dragging
+            endDragging()
+        }
+        super.touchesBegan(touches, withEvent: event)
+    }
+
+    override func touchesMoved(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if previousScroll != nil {
+            if let touch = touches.first as? UITouch,
+                let tab = bvc?.tabManager.selectedTab,
+                let prev = self.previousScroll,
+                let startOffset = self.bvc?.startOffset {
+                    let offset = touch.locationInView(bvc?.view)
+                    let dy = offset.y - prev.y
+
+                    if !tab.loading &&
+                        offset.y > -startOffset  &&
+                        offset.y < bvc!.tabManager.selectedTab!.webView.scrollView.contentSize.height {
+                            self.bvc?.scrollFooter(dy)
+                            self.bvc?.scrollHeader(dy)
+                            self.bvc?.scrollReader(dy)
+                    }
+                    self.previousScroll = offset
+            }
         }
+        super.touchesMoved(touches, withEvent: event)
     }
 
-    func scrollViewWillEndDragging(scrollView: UIScrollView, withVelocity velocity: CGPoint,
-        targetContentOffset: UnsafeMutablePointer<CGPoint>) {
-        self.previousScroll = nil
+    override func touchesCancelled(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        super.touchesCancelled(touches, withEvent: event)
+    }
 
-        if shouldAnimateToolbarsOpened() {
-            showToolbars(animated: true)
-        } else if shouldAnimateToolbarsClosed() {
-            hideToolbars(animated: true)
+    override func touchesEnded(touches: Set<NSObject>!, withEvent event: UIEvent!) {
+        if previousScroll != nil && touches.count == 1 {
+            endDragging()
+            self.previousScroll = nil
```

Comment:
Don't think you need this since we're nil'ing out self.previousScroll in endDragging()